core/src/test/scala/org/apache/spark/sql/test/generator/TestDataGenerator.scala

```diff
@@ -0,0 +1,375 @@
+
+    val idxColumns: Map[String, (List[(String, Integer)], Boolean)] =
+      singleColumnPrimaryKey ++
+        indices.map { idx =>
+          val columns = idx.indexColumns.map(x => (columnNames(x.getId), x.getLength))
+          if (idx.isPrimaryKey && idx.indexColumns.size == 1) {
+            val pkColumn = columnNames(idx.indexColumns.head.getId)
+            val x = columnDesc(pkColumn)
+            columnDesc(pkColumn) = (x._1, x._2, x._3 + " not null")
+          }
+          (generateIndexName(columns.map { _._1 }), (columns, idx.isPrimaryKey))
+        }.toMap
+
+    Schema(database, table, columnNames, columnDesc.toMap, idxColumns)
+  }
+
+  private def generateRandomValue(row: TiRow,
+                                  offset: Int,
+                                  r: Random,
+                                  valueGenerator: ValueGenerator): Unit = {
+    if (valueGenerator.randomNull(r)) {
+      row.setNull(offset)
+    } else {
+      val value = valueGenerator.randomValue(r)
+      row.set(offset, valueGenerator.tiDataType, value)
+    }
+  }
+
+  private def generateRandomRows(schema: Schema, n: Long, r: Random): List[TiRow] = {
+    (1.toLong to n).map { _ =>
+      val length = schema.columnInfo.length
+      val row: TiRow = ObjectRowImpl.create(length)
+      for (i <- schema.columnInfo.indices) {
+        val columnInfo = schema.columnInfo(i)
+        generateRandomValue(row, i, r, columnInfo.generator)
+      }
+      row
+    }.toList
+  }
+
+  // Data Generator
+  def randomDataGenerator(schema: Schema, rowCount: Long, directory: String, r: Random): Data = {
+    Data(schema, generateRandomRows(schema, rowCount, r), directory)
+  }
+}
+
+class TestDataGenerator extends SparkFunSuite {
+
+  test("base test for schema generator") {
+    val r = new Random(1234)
+    val schema = TestDataGenerator.schemaGenerator(
+      "tispark_test",
+      "test_table",
+      r,
+      List(
+        (INT, "", "not null primary key"),
+        (INT, "", "default null"),
+        (DOUBLE, "", "not null default 0.2"),
+        (VARCHAR, "50", "default null"),
+        (DECIMAL, "20,3", "default null")
+      ),
+      List(
+        Key(List(DefaultColumn(2), DefaultColumn(3))),
+        Key(List(PrefixColumn(4, 20))),
+        Key(List(DefaultColumn(3), DefaultColumn(5)))
+      )
+    )
+    val schema2 = TestDataGenerator.schemaGenerator(
+      "tispark_test",
+      "test_table",
+      r,
+      List(
+        (INT, "", "not null"),
+        (INT, "", "default null"),
+        (DOUBLE, "", "not null default 0.2"),
+        (VARCHAR, "50", "default null"),
+        (DECIMAL, "20,3", "default null")
+      ),
+      List(
+        PrimaryKey(List(DefaultColumn(1))),
+        Key(List(DefaultColumn(2), DefaultColumn(3))),
+        Key(List(PrefixColumn(4, 20))),
+        Key(List(DefaultColumn(3), DefaultColumn(5)))
+      )
+    )
+    println(schema)
```

Comment:
remove this println

Reply:
it is actually a hint for developer to acknowledge what test generator really output in this test. I think it can be left unchanged.

Reply:
okay